### PR TITLE
Fix OverflowError on Celery worker

### PR DESCRIPTION
**Pull request by guyco33**
## What type of PR is this? (check all applicable)

- [x] Bug Fix

## Description
`next_iteration += datetime.timedelta(minutes=2**failures)` start failing when number of failures is larger then 31 and celery worker stops refreshing queries

### CODE REVIEW

File: redash/models/__init__.py [hunk 1]
```diff
@@ -354,7 +354,10 @@ def should_schedule_next(previous_iteration, now, interval, time=None, day_of_we
         next_iteration = (previous_iteration + datetime.timedelta(days=days_delay) +
                           datetime.timedelta(days=days_to_add)).replace(hour=hour, minute=minute)
     if failures:
-        next_iteration += datetime.timedelta(minutes=2**failures)
+        try:
```
Comment (rauchy) [inline redash/models/__init__.py]: What you're saying is basically equivalent to "after failure 31, the next scheduled time is soooo far ahead in the future that there's really no point in scheduling it, meaning that the `should_schedule_next` function should simply return `False`, instead of a magic date in the post apocalyptic world. 😄 

How about this?

```suggestion
        try:
            next_iteration += datetime.timedelta(minutes=2**failures)
            return now > next_iteration
        except OverflowError:
            return False            
```

Comment (guyco33) [reply redash/models/__init__.py]: Thanks. Make sense

Comment (rauchy) [reply redash/models/__init__.py]: Seems like the suggestion wasn't applied correctly. Not sure Github is supporting multi-line suggestions ATM.

image

Comment (guyco33) [reply redash/models/__init__.py]: `return now > next_iteration` should be outside the `if failures` to cover also the non failure cases.
I fixed it accordingly

File: tests/test_models.py [hunk 2]
```diff
@@ -130,7 +130,7 @@ def test_backoff(self):
                                                     failures=5))
         self.assertFalse(models.should_schedule_next(two_hours_ago, now,
                                                      "3600", failures=10))
-
+        self.assertFalse(models.should_schedule_next(two_hours_ago, now, "3600", failures=32))
```
Comment (rauchy) [inline tests/test_models.py]: @guyco33 sorry for being a pain in the ass about this, but I'm a firm believer in one assert per test, and specifically, one logical assert per test.

The two assertions above the one you added are kind of related - they test the flip side of the same conditional, and belong in the same test (although this is debatable). The assertion you added feels like it tests a different concept, and I believe it should be separated.

Also - in the absence self-explanatory code, it's the tests' job to provide documentation, and just asserting "false" when failures=32 feels like a magic number with magic behavior. A dedicated test, with an explanatory test name, would really help here.

Comment (guyco33) [reply tests/test_models.py]: @rauchy I totally agree with you, especially when the tests cover special behaviour of magic numbers and unique corner cases.  
I will pick a proper name for a new assertion and will update the PR accordingly

**Comment by rauchy** [comment]
Thanks @guyco33!
